### PR TITLE
feat: install multiple subsets via command line

### PR DESCRIPTION
**Pull request by ispyhumanfly**
We love this module. And need it for our multi-platform environments. We found the need to be able to make it so "subset install input_string" could handle multiple input strings. Allowing one single invocation to process numerous subsets at once. We also made, what we feel, are some very small optimizations to the use of commander to handle this case. Updated the README to reflect these changes. And tested to make sure that this would not break other peoples current usage of the tool.

### CODE REVIEW

**Comment by ispyhumanfly** [comment]
We've been using this pretty frequently in our Docker containers without great success.

**Comment by ispyhumanfly** [comment]
Hey @tabrindle, we've been using our version of install-subset in production for a while now and it's really working nicely. Wanted to know your thoughts on the additions we've made and if you'd like to take them into your mainline.

**Comment by tabrindle** [comment]
Apologies for taking a long time to get to this. Saw the original notification and it got buried :-( So yes, docker containers is a huge use case here, and I think you are definitely onto something - however I have some hesitations about the implementation here. 

The original API has a mutually exclusive option of either include or exclude. (the if/else near pick/omit) Having multiple possible subsets to install at the same time breaks this pattern, and opens us up to to having infinite possible combinations of include and exclude. For example, one subset could be all include, and one could be all exclude. The current implementation doesn't seem to account for this. While we could resolve these, we would have to make some code changes, though I think the end effect would be more confusion than anything. 

However, I agree there is definitely something missing here. There is no reason to run install multiple times for different subsets - so what I propose instead is adding cosmiconfig (something I have been meaning to do anyways) to the project, so that we can have an external config file (outside of package.json) capable of doing this merge/pick/omit in user land. 

Instead of having multiple subsets in the cli command, we merge (intersection, xor, pick, omit, etc) them in the config and give it a new name, to be called by itself. I believe this achieves your purpose, as well as gives us a few other features, and the ability to clean up the package.json for complex or large projects.

subset.config.js
```
const merge = require('lodash/merge');

const sync = {
  include: [
    'lodash',
    'fibrous',
    'node-fetch',
    'npm-registry-client',
  ],
};

const test = {
  include: [
    'babel',
    'jest',
    'jest-junit',
  ]
}

module.exports = {
  subsets: {
    sync,
    test,
    both: merge(test, sync)
  },
},
```

`$ npx install-subset install both`

There's nothing saying we can't implement both your feature and the above, however, I would like to have a warning to prevent the both include/exclude footgun.

File: index.js [hunk 1]
```diff
@@ -40,34 +40,56 @@ var pick = function(obj, props) {
...
 
     if (!packageJson.subsets) {
-      throw 'No install subsets in package.json';
+      console.error('No install subsets in package.json.')
+      process.exit(1)
     }
 
-    if (!packageJson.subsets[input_string]) {
-      throw 'No install subset with that name';
+    const subsetStrings = new Array()
```
Comment (tabrindle) [inline index.js]: while there's no difference here, its a little more conventional to use an array literal []

File: index.js [hunk 2]
```diff
@@ -40,34 +40,56 @@ var pick = function(obj, props) {
...
+  .option('-d, --clean', 'Remove node_modules first.')
+  .option('--npm', 'Use npm to install.')
+  .option('-a, --all', 'Prune dependencies as well as devDependencies.')
+  .description('Install a given subset or multiple subsets defined in package.json.')
+  .action(function(input_strings, options) {
 
     if (!packageJson.subsets) {
-      throw 'No install subsets in package.json';
+      console.error('No install subsets in package.json.')
+      process.exit(1)
```
Comment (tabrindle) [inline index.js:53]: I like this better 👍

File: index.js [hunk 3]
```diff
@@ -108,10 +131,14 @@ cli
...
     if (installer.status !== 0) {
-      throw 'Error code ' + installer.status;
+
+      console.error('Error code ' + installer.status + '.')
+      process.exit(1)
     }
 
-    console.log('Installation of subset "' + input_string + '" successful');
+    for (let input_string of subsetStrings) {
+      console.log('Installation of subset "' + input_string + '" successful.');
```
Comment (tabrindle) [inline index.js]: probably best to use a template`${input_string}` here

File: index.js [hunk 4]
```diff
@@ -40,34 +40,56 @@ var pick = function(obj, props) {
...
+        }
+        subset.exclude = [...subset.exclude, ...packageSubset.exclude]
+      }
+      if (packageSubset.include) {
+        if (!subset.include) {
+          subset.include = []
+        }
+        subset.include = [...subset.include, ...packageSubset.include]
+      }
+    }
```
Comment (tabrindle) [inline index.js:83]: Re: my comment in the PR description thread - we should at least have a warning here, possibly an exit condition if both an exclude and include are in the subset object.

**Review by tabrindle — CHANGES_REQUESTED** [review]

File: index.js (hunk 2 above)
Comment (ispyhumanfly) [reply index.js:53]: Sweet :)

File: index.js (hunk 1 above)
Comment (ispyhumanfly) [reply index.js:56]: It's more old school. I'm down with that, too.

**Comment by Logikgate** [comment]
@tabrindle thanks for the review! We have made the code style changes and made it exit with an error code if the subset or combination of subsets contains both includes and excludes.  Let us know what you think

File: index.js [hunk 5]
```diff
@@ -40,34 +40,62 @@ var pick = function(obj, props) {
...
+        if (!subset.include) {
+          subset.include = []
+        }
+        subset.include = [...subset.include, ...packageSubset.include]
+      }
     }
 
-    const subset = packageJson.subsets[input_string];
+    if (subset.include && subset.exclude) {
+        console.erro(`Failed to install subset ${input_string}.`);
```
Comment (tabrindle) [inline index.js]: typo `console.erro`

File: index.js (hunk 4 above)
Comment (ispyhumanfly) [reply index.js:83]: Agreed

**Comment by tabrindle** [comment]
Changes look good, fix the last typo and I think we are good to merge. Then I'll merge master into my PR and we'll be good to make a new release!

**Review by tabrindle — CHANGES_REQUESTED** [review]

**Comment by ispyhumanfly** [comment]
> Changes look good, fix the last typo and I think we are good to merge. Then I'll merge master into my PR and we'll be good to make a new release!

Which typo was that? I'll fix it now.

**Comment by tabrindle** [comment]
console.erro(`Failed to install subset ${input_string}.`);
=>
console.error

File: index.js (hunk 5 above)
Comment (ispyhumanfly) [reply index.js]: This is now resolved.

**Comment by ispyhumanfly** [comment]
> console.erro(`Failed to install subset ${input_string}.`);
> =>
> console.error

Ahh, yes... I didn't catch that. But I did resolve that yesterday and pushed it into master.

**Pull request merged**